lib/Interpreter/CIFactory.cpp

```diff
@@ -928,29 +961,132 @@ namespace {
   }
 
   static void AddRuntimeIncludePaths(llvm::StringRef ClingBin,
-                                     clang::HeaderSearchOptions& HOpts) {
-    if (HOpts.Verbose)
-      llvm::errs() << "Adding runtime include paths:\n";
-    // Add configuration paths to interpreter's include files.
+                                    std::vector<const char*>& args,
+                                    bool Verbose) {
+    static std::vector<std::string> sArguments;
+    if (sArguments.empty()) {
 #ifdef CLING_INCLUDE_PATHS
-    if (HOpts.Verbose)
-      llvm::errs() << "  \"" CLING_INCLUDE_PATHS "\"\n";
-    utils::AddIncludePaths(CLING_INCLUDE_PATHS, HOpts);
+      llvm::SmallVector<llvm::StringRef, 8> Paths;
+      cling::utils::SplitPaths(CLING_INCLUDE_PATHS, Paths);
+      for (llvm::StringRef& Path : Paths)
+        sArguments.push_back(Path.str());
 #endif
-    llvm::SmallString<512> P(ClingBin);
-    if (!P.empty()) {
-      // Remove /cling from foo/bin/clang
-      llvm::StringRef ExeIncl = llvm::sys::path::parent_path(P);
-      // Remove /bin   from foo/bin
-      ExeIncl = llvm::sys::path::parent_path(ExeIncl);
-      P.resize(ExeIncl.size());
-      // Get foo/include
-      llvm::sys::path::append(P, "include");
-      if (llvm::sys::fs::is_directory(P.str()))
-        utils::AddIncludePaths(P.str(), HOpts, nullptr);
+      llvm::SmallString<1024> P(ClingBin);
+      if (!P.empty()) {
+        // Remove /cling from foo/bin/clang
+        llvm::StringRef ExeIncl = llvm::sys::path::parent_path(P);
+        // Remove /bin   from foo/bin
+        ExeIncl = llvm::sys::path::parent_path(ExeIncl);
+        P.resize(ExeIncl.size());
+        // Get foo/include
+        llvm::sys::path::append(P, "include");
+        if (llvm::sys::fs::is_directory(P))
+          sArguments.push_back(P.str());
+      }
     }
+    for (const std::string& arg : sArguments) {
+      args.push_back("-I");
+      args.push_back(arg.c_str());
+    }
+
+    if (Verbose) {
+      llvm::errs() << "Added runtime include paths:\n";
+      for (const std::string& arg : sArguments)
+        llvm::errs() << "  " << arg << "\n";
+    }
+  }
+
+  static llvm::IntrusiveRefCntPtr<DiagnosticsEngine>
```

Comment:
Could you implement that without statics? Statics are troublesome for multi interpreter setups.